text/0000-named-custom-cargo-profiles.md

```diff
@@ -0,0 +1,275 @@
+    lto = true
+
+Passing `--profile` with the profile's name to various Cargo commands will
+resolve to the custom profile. Overrides specified in the profiles from which
+the custom profile inherits will be inherited too, and all final outputs may
+go to a different directory by default:
+
+    $ cargo build
+    $ cargo build --release
+    $ cargo build --profile release-lto
+    $ ls -l target
+    debug release release-lto
+
+Cargo will emit errors in case `inherits` loops are detected. When considering
+inheritence hierarchy, all profiles directly or indirectly inherit from
+either from `release` or from `dev`.
+
+This also affects other Cargo commands:
+
+* `cargo test` also receives `--profile`, but unless it is specified, uses
+  the predefined `test` profile which is described below.
+* `cargo bench` also receives `--profile`, but unless it is specified, uses
+  the predefined `bench` profile which is described below.
+* `cargo doc` also receives `--profile`, but unless it is specified, uses
+  the predefined `doc` profile which is described below.
+
+
+## New `dir-name` attribute
+
+Some of the paths generated under `target/` have resulted in a de-facto "build
+protocol", where `cargo` is invoked as a part of a larger project build. So, to
+preserve the existing behavior, there is also a new attribute `dir-name`, which
+when left unspecified, defaults to the name of the profile. For example:
+
+    [profile.release-lto]
+    inherits = "release"
+	dir-name = "lto"  # Emits to target/lto instead of target/release-lto
+    lto = true
+
+* The `dir-name` attribute is used mainly to direct the outputs of `bench` and
+  `test` to their respective directories: `target/release` and `target/debug`.
+  This preserves existing behavior.
+* The `dir-name` attribute is the only attribute not passed by inheritance.
+* Path seperators are not allowed in `dir-name`.
+
+
+## Treatment to the pre-defined profiles
+
+* The `release` profile remains as it is, with settings overridable as
+  before.
+* The `dev` profile receives the `dir-name = "debug"` attribute, so that its
+  final outputs are emitted to `target/debug`, as existing Rust developers will
+  expect this behavior. This should be added in the official
+  documentation for the Cargo manifest, to make this fact clearer for users.
+* A `debug` profile name is not allowed, with a warning saying to use the
+  already established `dev` name.
+* The `bench` profile defaults to the following definition, to preserve current
+  behavior:
+
+```
+[profile.bench]
+inherits = "release"
+dir-name = "release"
+```
+
+* The `test` profile defaults to the following definition, to preserve current behavior:
+```
+[profile.test]
+inherits = "dev"
+dir-name = "debug"
+```
+
+* The `doc` profile defaults to the following definition, to preserve current
+  behavior:
+```
+[profile.doc]
+inherits = "dev"
+dir-name = "debug"
+```
+
+
+* The (upcoming) `build` profile defaults to the following definition:
```

Comment:
Maybe add a note that this is experimental/undecided.

Reply:
Sure.